processing/src/main/java/io/druid/query/extraction/namespace/URIExtractionNamespace.java

```diff
@@ -0,0 +1,508 @@
+      URI uri,
+      @JsonProperty(value = "namespaceParseSpec", required = true)
+      FlatDataParser namespaceParseSpec,
+      @Min(0) @Nullable @JsonProperty(value = "pollPeriod", required = false)
+      Period pollPeriod,
+      @JsonProperty(value = "versionRegex", required = false)
+      String versionRegex
+  )
+  {
+    if (versionRegex != null) {
+      try {
+        Pattern.compile(versionRegex);
+      }
+      catch (PatternSyntaxException ex) {
+        throw new IAE(ex, "Could not parse `versionRegex` [%s]", versionRegex);
+      }
+    }
+    this.namespace = Preconditions.checkNotNull(namespace, "namespace");
+    this.uri = Preconditions.checkNotNull(uri, "uri");
+    this.namespaceParseSpec = Preconditions.checkNotNull(namespaceParseSpec, "namespaceParseSpec");
+    this.pollPeriod = pollPeriod == null ? Period.ZERO : pollPeriod;
+
+    this.versionRegex = versionRegex;
+  }
+
+  @Override
+  public String getNamespace()
+  {
+    return namespace;
+  }
+
+  public String getVersionRegex()
+  {
+    return versionRegex;
+  }
+
+  public FlatDataParser getNamespaceParseSpec()
+  {
+    return this.namespaceParseSpec;
+  }
+
+  public URI getUri()
+  {
+    return uri;
+  }
+
+  @Override
+  public long getPollMs()
+  {
+    return pollPeriod.toStandardDuration().getMillis();
+  }
+
+  @Override
+  public String toString()
+  {
+    return String.format(
+        "URIExtractionNamespace = { namespace = %s, uri = %s, namespaceParseSpec = %s, pollPeriod = %s }",
+        namespace,
+        uri.toString(),
+        namespaceParseSpec.toString(),
+        pollPeriod.toString()
+    );
+  }
+
+
+  @Override
+  public boolean equals(Object o)
+  {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    URIExtractionNamespace namespace1 = (URIExtractionNamespace) o;
+    return toString().equals(namespace1.toString());
+  }
+
+  @Override
+  public int hashCode()
+  {
+    int result = namespace.hashCode();
+    result = 31 * result + uri.hashCode();
+    result = 31 * result + namespaceParseSpec.hashCode();
+    result = 31 * result + pollPeriod.hashCode();
+    result = 31 * result + (versionRegex != null ? versionRegex.hashCode() : 0);
+    return result;
+  }
+
+
+  private static class DelegateParser implements Parser<String, String>
+  {
+    private final Parser<String, Object> delegate;
+    private final String key;
+    private final String value;
+
+    private DelegateParser(
+        Parser<String, Object> delegate,
+        String key,
+        String value
+    )
+    {
+      this.delegate = delegate;
+      this.key = key;
+      this.value = value;
+    }
+
+    @Override
+    public Map<String, String> parse(String input)
+    {
+      final Map<String, Object> inner = delegate.parse(input);
+      final String k = (String) Preconditions.checkNotNull(
+          inner.get(key),
```

Comment:
It should never be null here, I'm moving things around in the constructors a bit to ensure that